### PR TITLE
impl(rest): add RestContext

### PR DESCRIPTION
**Pull request by scotthart**
Part of the work for #10037

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038)

### CODE REVIEW

File: google/cloud/internal/rest_context.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  // Header names are case-insensitive; header values are case-sensitive.
+  std::vector<std::string> GetHeader(std::string header) const;
+
+ private:
+  friend bool operator==(RestContext const& lhs, RestContext const& rhs);
+  HttpHeaders headers_;
+};
+
+bool operator==(RestContext const& lhs, RestContext const& rhs);
+bool operator!=(RestContext const& lhs, RestContext const& rhs);
```
Comment (coryan) [inline google/cloud/internal/rest_context.h:63]: Don't we need to make this a friend too?

File: google/cloud/internal/rest_context.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_REST_CONTEXT_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_REST_CONTEXT_H
+
+#include "google/cloud/status_or.h"
```
Comment (coryan) [inline google/cloud/internal/rest_context.h:18]: unused?

File: google/cloud/internal/rest_context.cc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+RestContext& RestContext::AddHeader(
+    std::pair<std::string, std::string> header) & {
+  return AddHeader(std::move(header.first), std::move(header.second));
+}
+
+std::vector<std::string> RestContext::GetHeader(std::string header) const {
+  std::transform(header.begin(), header.end(), header.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
```
Comment (coryan) [inline google/cloud/internal/rest_context.cc:44]: `std::string::value_type` is `char`, why accept `unsigned char` here?

Comment (coryan) [reply google/cloud/internal/rest_context.cc:44]: `auto` cures these ills btw.

File: google/cloud/internal/rest_context_test.cc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  auto context = RestContext()
+                     .AddHeader("header1", "value1")
+                     .AddHeader(std::make_pair("header2", "value2a"))
+                     .AddHeader("header2", "value2b");
+  EXPECT_THAT(
+      context.headers(),
+      Contains(std::make_pair("header1", std::vector<std::string>{"value1"})));
+  EXPECT_THAT(context.headers(),
+              Contains(std::make_pair(
+                  "header2", std::vector<std::string>{"value2a", "value2b"})));
```
Comment (coryan) [inline google/cloud/internal/rest_context_test.cc:48]: Maybe `Contains(Pair("header2", ElementsAre("value2a", "value2b")))` ?

**Review by coryan — APPROVED** [review]
A few nits, but LGTM

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038)* status: 0 of 8 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/internal/rest_context.h` line 18 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038#-NECFofIsmj0Yp6Yijo4-r1-18:-NECdpoz1WOCwtruFmzv:ba3dfob) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/01b62a106b815a782372480f9b40dd04639c8217/google/cloud/internal/rest_context.h#L18)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Removed
___
*[`google/cloud/internal/rest_context.h` line 63 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038#-NECFofIsmj0Yp6Yijo4-r1-63:-NECdqdf4ub2tK6qNYyd:b-sk63tm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/01b62a106b815a782372480f9b40dd04639c8217/google/cloud/internal/rest_context.h#L63)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Don't we need to make this a friend too?
</blockquote></details>

operator!= calls operator==, so only operator== needs to be a friend
___
*[`google/cloud/internal/rest_context.cc` line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038#-NECFofIsmj0Yp6Yijo3-r1-44:-NECdxcr9YNP-lZYCqv0:b-f447x0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/01b62a106b815a782372480f9b40dd04639c8217/google/cloud/internal/rest_context.cc#L44)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

`auto` cures these ills btw.
</blockquote></details>

switched to auto
___
*[`google/cloud/internal/rest_context_test.cc` line 48 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10038#-NECFofIsmj0Yp6Yijo5-r1-48:-NECe-Kp24TLTVPpzDCd:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/01b62a106b815a782372480f9b40dd04639c8217/google/cloud/internal/rest_context_test.cc#L48)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe `Contains(Pair("header2", ElementsAre("value2a", "value2b")))` ?
</blockquote></details>

Done.

File: google/cloud/internal/rest_context.cc [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+RestContext& RestContext::AddHeader(std::string header, std::string value) & {
+  std::transform(header.begin(), header.end(), header.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+  auto iter = headers_.find(header);
+  if (iter == headers_.end()) {
+    std::vector<std::string> v = {std::move(value)};
+    headers_.emplace(std::move(header), std::move(v));
+  } else {
+    iter->second.push_back(value);
```
Comment (devbww) [inline google/cloud/internal/rest_context.cc:32]: `std::move(value)`

File: google/cloud/internal/rest_context.cc [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+RestContext& RestContext::AddHeader(
+    std::pair<std::string, std::string> header) & {
+  return AddHeader(std::move(header.first), std::move(header.second));
+}
+
+std::vector<std::string> RestContext::GetHeader(std::string header) const {
+  std::transform(header.begin(), header.end(), header.begin(),
+                 [](auto c) { return std::tolower(c); });
```
Comment (devbww) [inline google/cloud/internal/rest_context.cc:44]: It is important that the `tolower()` argument is first converted to an `unsigned char` (before then being coerced to an `int`), just like you have in the `AddHeader()` case.  Indeed, there is an exact example of this on the `std::tolower` cppreference page.

Alternatively, you could use `absl::AsciiStrToLower(&header)`.

Comment (coryan) [reply google/cloud/internal/rest_context.cc:44]: TIL: https://github.com/googleapis/google-cloud-cpp/pull/10041 to fix this everywhere.

**Pull request merged**